File: 프로그래머스/lv3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    vector<bool> visited(n, false);
    
    for (int i = 0; i < n; i++) {
        
        if (visited[i])
            continue;
        
        queue<int> q;
        q.push(i);
        
        while (!q.empty()) {
            int v = q.front();
            q.pop();
            
            if (visited[v])
                continue;
            
            visited[v] = true;
            
            for (int j = 0; j < n; j++) {
                if (v != j && computers[v][j] == 1 && !visited[j])
                    q.push(j);
            }
        }
        
        answer++;
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    visited = [False] * n
    
    def dfs(num):
        visited[num] = True
        for neighbor in range(n):
            if visited[neighbor] == False and computers[num][neighbor] == 1:
                dfs(neighbor)
    
    for i in range(n):
        if not visited[i]:
            dfs(i)
            answer += 1
    return answer

